client/disputes/test/index.js

```diff
@@ -24,53 +82,137 @@ describe( 'Disputes list', () => {
 	test( 'renders correctly', () => {
 		useDisputes.mockReturnValue( {
 			isLoading: false,
-			disputes: [
-				{
-					id: 'dp_asdfghjkl',
-					amount: 1000,
-					currency: 'usd',
-					created: 1572590800,
-					evidence_details: {
-						due_by: 1573199200,
-					},
-					reason: 'fraudulent',
-					status: 'needs_response',
-					charge: {
-						id: 'ch_mock',
-						payment_method_details: {
-							card: {
-								brand: 'visa',
-							},
-						},
-						billing_details: {
-							name: 'Mock customer',
-							email: 'mock@customer.net',
-							address: {
-								country: 'US',
-							},
-						},
-					},
-					order: {
-						number: '1',
-						url: 'http://test.local/order/1',
-					},
-				},
-				{
-					id: 'dp_zxcvbnm',
-					amount: 1050,
-					currency: 'usd',
-					created: 1572480800,
-					evidence_details: {
-						due_by: 1573099200,
-					},
-					reason: 'general',
-					status: 'under_review',
-					// dispute without order or charge information
-				},
-			],
+			disputes: mockDisputes,
 		} );
 
 		const { container: list } = render( <DisputesList /> );
 		expect( list ).toMatchSnapshot();
 	} );
+
+	describe( 'Download button', () => {
+		test( 'renders when there are one or more disputes', () => {
+			useDisputes.mockReturnValue( {
+				disputes: mockDisputes,
+				isLoading: false,
+			} );
+
+			const { queryByRole } = render( <DisputesList /> );
+			const button = queryByRole( 'button', { name: 'Download' } );
+
+			expect( button ).not.toBeNull();
+		} );
+
+		test( 'does not render when there are no disputes', () => {
+			useDisputes.mockReturnValue( {
+				disputes: [],
+				isLoading: false,
+			} );
+
+			const { queryByRole } = render( <DisputesList /> );
+			const button = queryByRole( 'button', { name: 'Download' } );
+
+			expect( button ).toBeNull();
+		} );
+	} );
+
+	describe( 'CSV download', () => {
+		beforeEach( () => {
+			useDisputes.mockReturnValue( {
+				disputes: mockDisputes,
+				isLoading: false,
+			} );
+		} );
+
+		afterEach( () => {
+			jest.resetAllMocks();
+		} );
+
+		afterAll( () => {
+			jest.restoreAllMocks();
+		} );
+
+		test( 'should render expected columns in CSV when the download button is clicked ', () => {
+			const { getByRole } = render( <DisputesList /> );
+			getByRole( 'button', { name: 'Download' } ).click();
+
+			const expected = [
+				'"Dispute Id"',
+				'Amount',
+				'Status',
+				'Reason',
+				'Source',
+				'"Order #"',
+				'Customer',
+				'Email',
+				'Country',
+				'"Disputed on"',
+				'"Respond by"',
+			];
+
+			const csvContent = downloadCSVFile.mock.calls[ 0 ][ 1 ];
+			const csvHeaderRow = csvContent.split( os.EOL )[ 0 ].split( ',' );
+			expect( csvHeaderRow ).toEqual( expected );
+		} );
+
+		test( 'should match the visible rows', () => {
```

Comment:
In my original PR, I made the code in this test deliberately verbose for readability. If you can think of clever tweaks without losing readability, that'd be awesome.

Reply:
Thanks for checking it @anu-rock. The only thing I would do differently is to base the assertions in the data contained in the mockDisputes rather than in the text content of the rendered table. If I get this right, the csv file is generated based on the data that comes from the API, not from the data that is in the table.

Maybe testing against the data makes it a bit easier to test, I am going to give it a try.

Reply:
@anu-rock I changed the approach, following my previous comment. It feels more readable to me know, what do you think?

Reply:
If I'm not wrong, there's a fair amount of processing/formatting of data involved between using API data and displaying it in UI:

https://github.com/Automattic/woocommerce-payments/blob/fc4f4449151ea0732501ee29147a793088b3c9a4/client/disputes/index.js#L105-L198

So, what we export is the displayed data rather than the raw data from the API response. To me, it makes sense to validate CSV data against displayed data.

Reply:
@anu-rock fair enough. I reverted to use previous approach.

Following one of your previous comments I also moved the utility functions to a common test utils file